src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,383 @@
+from py42 import exceptions
+from py42.exceptions import Py42NotFoundError
+
+from code42cli.bulk import run_bulk_process
+from code42cli.click_ext.groups import OrderedGroup
+from code42cli.errors import Code42CLIError
+from code42cli.file_readers import read_csv_arg
+from code42cli.options import format_option
+from code42cli.options import sdk_options
+from code42cli.output_formats import DataFrameOutputFormatter
+from code42cli.output_formats import OutputFormatter
+
+
+@click.group(cls=OrderedGroup)
+@sdk_options(hidden=True)
+def devices(state):
+    """For managing devices within your Code42 environment."""
+    pass
+
+
+device_guid_argument = click.argument("device-guid", type=str)
+
+change_device_name_option = click.option(
+    "--change-device-name",
+    required=False,
+    is_flag=True,
+    default=False,
+    help="""Prepend "deactivated_" and today's date to the name of any
+    deactivated devices.""",
+)
+
+purge_date_option = click.option(
+    "--purge-date",
+    required=False,
+    type=str,
+    default=None,
+    help="""The date on which the archive should be purged from cold storage in yyyy-MM-dd format.
+    If not provided, the date will be set according to the appropriate org settings.""",
+)
+
+
+@devices.command()
+@device_guid_argument
+@change_device_name_option
+@purge_date_option
+@sdk_options()
+def deactivate(state, device_guid, change_device_name, purge_date):
+    """Deactivate a device within Code42. Requires the device GUID to deactivate."""
+    _deactivate_device(state.sdk, device_guid, change_device_name, purge_date)
+
+
+def _deactivate_device(sdk, device_guid, change_device_name, purge_date):
+    device = sdk.devices.get_by_guid(device_guid)
+    try:
+        sdk.devices.deactivate(device.data["computerId"])
+    except exceptions.Py42BadRequestError:
+        raise Code42CLIError("The device {} is in legal hold.".format(device_guid))
+    except exceptions.Py42NotFoundError:
+        raise Code42CLIError("The device {} was not found.".format(device_guid))
+    except exceptions.Py42ForbiddenError:
+        raise Code42CLIError("Unable to deactivate {}.".format(device_guid))
+    if purge_date or change_device_name:
+        guid = device_guid
+        name = device.data["name"]
+    if purge_date:
+        _update_cold_storage_purge_date(sdk, guid, purge_date)
+    if change_device_name and not name.startswith("deactivated_"):
+        _change_device_name(
+            sdk,
+            device_guid,
+            "deactivated_" + date.today().strftime("%Y-%m-%d") + "_" + name,
+        )
+
+
+def _update_cold_storage_purge_date(sdk, guid, purge_date):
+    archives_response = sdk.archive.get_all_by_device_guid(guid)
+    archive_guid_list = [
+        archive["archiveGuid"]
+        for page in archives_response
+        for archive in page["archives"]
+    ]
+    for archive_guid in archive_guid_list:
+        sdk.archive.update_cold_storage_purge_date(archive_guid, purge_date)
```

Comment:
This fails on v2 (security plan archives) since if the user is still active v2 archives will stay active and not have a purge date. We should filter those where `'format': 'ARCHIVE_V2'`.